### PR TITLE
RavenDB-13765: Recover datafile into recovered database

### PR DESCRIPTION
**Pull request by aviviadi**
As first stage, this code recovers successfully documents, attachments, revisions, and counters. It is also recovering conflicts (however it doesn't considering order). Need to add times series.

After that need to remove old exported dump files.

### CODE REVIEW

File: test/SlowTests/Issues/RavenDB-8451.cs [hunk 1]
```diff
@@ -54,7 +56,7 @@ public async Task CanRecoverEncryptedDatabaseInternal(bool nullifyMasterKey = fa
             var recoveryExportPath = NewDataPath(prefix: Guid.NewGuid().ToString());
 
             DatabaseStatistics databaseStatistics;
-
+            string recoverDbName = $"RecoverDB_{Guid.NewGuid().ToString()}";
```
Comment (ppekrol) [inline test/SlowTests/Issues/RavenDB-8451.cs:59]: wrap this in EnsureDatabaseDeletion

File: test/SlowTests/Issues/RavenDB_11904.cs [hunk 2]
```diff
@@ -27,26 +29,30 @@ public async Task CanLoadDatabaseAfterUsingVoronRecoveryOnItWithCopyOnWriteMode(
 
             DatabaseStatistics databaseStatistics;
 
+            string recoverDbName = $"RecoverDB_{Guid.NewGuid().ToString()}";
```
Comment (ppekrol) [inline test/SlowTests/Issues/RavenDB_11904.cs:31]: wrap this in EnsureDatabaseDeletion

File: test/SlowTests/Issues/RavenDB_11905.cs [hunk 3]
```diff
@@ -23,21 +25,24 @@ public async Task CanUseVoronRecoveryOnEmptyDatabase()
         {
             var dbPath = NewDataPath();
             var recoveryExportPath = NewDataPath();
-            
-            using (GetDocumentStore(new Options()
+            string recoverDbName = $"RecoverDB_{Guid.NewGuid().ToString()}";
```
Comment (ppekrol) [inline test/SlowTests/Issues/RavenDB_11905.cs:28]: wrap this in EnsureDatabaseDeletion

File: test/SlowTests/Voron/RecoveryTest.cs [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+            var flushed = false;
+            env.OnLogsApplied += () =>
+            {
+                flushed = true;
+            };
+            env.FlushLogToDataFile();
+            Assert.True(flushed, "Test requires successful flush to log");
+
+            string recoverDbName = $"RecoverDB_{Guid.NewGuid().ToString()}";
+            var _ = store.Maintenance.Server.Send(new CreateDatabaseOperation(new DatabaseRecord {DatabaseName = recoverDbName}));
```
Comment (ppekrol) [inline test/SlowTests/Voron/RecoveryTest.cs]: wrap this in EnsureDatabaseDeletion

File: tools/Voron.Recovery/RecoveredDatabaseCreator.cs [hunk 5]
```diff
@@ -0,0 +1,433 @@
...
+                {
+                    _database.DocumentsStorage.Put(_context, _logDocId, null, doc);
+                }
+                tx.Commit();
+            }
+        }
+
+        public void WriteDocument(Document document)
+        {
+            var actions = _databaseDestination.Documents();
```
Comment (ppekrol) [inline tools/Voron.Recovery/RecoveredDatabaseCreator.cs]: using

File: tools/Voron.Recovery/RecoveredDatabaseCreator.cs [hunk 6]
```diff
@@ -0,0 +1,433 @@
...
+        public void WriteRevision(Document document)
+        {
+            if (_database.DocumentsStorage.RevisionsStorage.Configuration == null)
+            {
+                var revisionsConfiguration = new RevisionsConfiguration {Default = new RevisionsCollectionConfiguration {Disabled = false}};
+                var configurationJson = EntityToBlittable.ConvertCommandToBlittable(revisionsConfiguration, _context);
+                (long index, _) = _database.ServerStore.ModifyDatabaseRevisions(_context, _database.Name, configurationJson, Guid.NewGuid().ToString()).Result;
+                AsyncHelpers.RunSync(() => _database.RachisLogIndexNotifications.WaitForIndexNotification(index, _database.ServerStore.Engine.OperationTimeout));
+            }
+            var actions = _databaseDestination.RevisionDocuments();
```
Comment (ppekrol) [inline tools/Voron.Recovery/RecoveredDatabaseCreator.cs]: using

File: tools/Voron.Recovery/RecoveredDatabaseCreator.cs [hunk 7]
```diff
@@ -0,0 +1,433 @@
...
+            }
+        }
+
+        public void WriteDocument(Document document)
+        {
+            var actions = _databaseDestination.Documents();
+            WriteDocumentInternal(document, actions);
+        }
+
+        public void WriteRevision(Document document)
```
Comment (ppekrol) [inline tools/Voron.Recovery/RecoveredDatabaseCreator.cs]: There should be no issue in changing this to async Task

File: tools/Voron.Recovery/RecoveredDatabaseCreator.cs [hunk 8]
```diff
@@ -0,0 +1,433 @@
...
+                using (var tx = _context.OpenWriteTransaction())
+                {
+                    Document originalDoc = _database.DocumentsStorage.Get(_context, counterGroupDetail.DocumentId);
+                    if (originalDoc == null)
+                    {
+                        var orphanDocId = $"{_orphanCountersDocIdPrefix}/{counterGroupDetail.DocumentId}";
+                        var orphanDoc = _database.DocumentsStorage.Get(_context, orphanDocId);
+                        if (orphanDoc == null)
+                        {
+                            using (var doc = _context.ReadObject(
```
Comment (ppekrol) [inline tools/Voron.Recovery/RecoveredDatabaseCreator.cs]: Wrap this in a method

File: tools/Voron.Recovery/RecoveredDatabaseCreator.cs [hunk 9]
```diff
@@ -0,0 +1,433 @@
...
+            _databaseDestination = new DatabaseDestination(documentDatabase);
+            _results = new SmugglerResult();
+            _databaseDestination.Initialize(new DatabaseSmugglerOptionsServerSide(), _results, ServerVersion.Build);
+            _contextDisposal = _database.DocumentsStorage.ContextPool.AllocateOperationContext(out _context);
+            _byteStringContext = new ByteStringContext(SharedMultipleUseFlag.None);
+            Slice.From(_byteStringContext, "Attachments", ByteStringType.Immutable, out _attachmentsSlice);
+
+            using (var tx = _context.OpenWriteTransaction())
+            {
+                using (var doc = _context.ReadObject(
```
Comment (ppekrol) [inline tools/Voron.Recovery/RecoveredDatabaseCreator.cs:61]: Wrap this in a method

File: tools/Voron.Recovery/RecoveredDatabaseCreator.cs [hunk 10]
```diff
@@ -0,0 +1,433 @@
...
+                            continue;
+                        if (!(attachmentDataObj is BlittableJsonReaderObject attachmentData))
+                            continue;
+
+                        var seenDoc = _database.DocumentsStorage.Get(_context, docId);
+                        if (seenDoc != null)
+                        {
+                            if (attachmentData.TryGet("Name", out string originalName) == false)
+                                originalName = name;
+                            if (attachmentData.TryGet("ContentType", out string originalContentType) == false)
```
Comment (ppekrol) [inline tools/Voron.Recovery/RecoveredDatabaseCreator.cs]: nameof

File: tools/Voron.Recovery/RecoveredDatabaseCreator.cs [hunk 11]
```diff
@@ -0,0 +1,433 @@
...
+                            continue;
+                        if (orphanAttachmentDoc.Data.TryGetMember(docId, out var attachmentDataObj) == false)
+                            continue;
+                        if (!(attachmentDataObj is BlittableJsonReaderObject attachmentData))
+                            continue;
+
+                        var seenDoc = _database.DocumentsStorage.Get(_context, docId);
+                        if (seenDoc != null)
+                        {
+                            if (attachmentData.TryGet("Name", out string originalName) == false)
```
Comment (ppekrol) [inline tools/Voron.Recovery/RecoveredDatabaseCreator.cs]: nameof

**Review by ppekrol — COMMENTED** [review]
General direction looks good to me. We need to focus on polishing and code organization.

**Comment by ayende** [comment]
What do you mean? " set affinity" ?

File: tools/Voron.Recovery/RecoveredDatabaseCreator.cs (hunk 5 above)
Comment (aviviadi) [reply tools/Voron.Recovery/RecoveredDatabaseCreator.cs]: no using here because sometimes I do want manually dispose it (and then enforce commit)

**Comment by aviviadi** [comment]
> What do you mean? " set affinity" 

I didn't understand the remark. To what place in code is this referring to?

**Comment by ppekrol** [comment]
> > What do you mean? " set affinity"
> 
> I didn't understand the remark. To what place in code is this referring to?

If refers to your comment at the top of the PR:
```
For now, Raven.voron executable needs linux capabilities to set affinity, and will have to be refactored.
```

File: tools/Voron.Recovery/CommandLineApp.cs [hunk 12]
```diff
@@ -67,16 +75,49 @@ private static void ConfigureRecoveryCommand()
                 var ignoreInvalidJournalErrorsArg = cmd.Option("--IgnoreInvalidJournalErrors", "Default is false.", CommandOptionType.SingleValue);
                 var ignoreDataIntegrityErrorsOfAlreadySyncedTransactionsArg = cmd.Option("--IgnoreInvalidDataErrorsOfAlreadySyncedTransactions", "Default is false.", CommandOptionType.SingleValue);
                 var ignoreInvalidPagesInARowArg = cmd.Option("--IgnoreInvalidPagesInARow", "Default is false.", CommandOptionType.SingleValue);
-
                 var loggingModeArg = cmd.Option("--LoggingMode", "Logging mode: Operations or Information.", CommandOptionType.SingleValue);
-
                 var masterKey = cmd.Option("--MasterKey", "Encryption key: base64 string of the encryption master key", CommandOptionType.SingleValue);
+                var recoveredServerPath = cmd.Option("--RecoveredServerPath", "Recovered database server destination", CommandOptionType.SingleValue);
```
Comment (ayende) [inline tools/Voron.Recovery/CommandLineApp.cs]: Why is this required? Shouldn't we just use `RecoverDirectory` ?

File: tools/Voron.Recovery/CommandLineApp.cs [hunk 13]
```diff
@@ -67,16 +75,49 @@ private static void ConfigureRecoveryCommand()
...
 
                 cmd.OnExecute(() =>
                 {
+                    if (recoveredServerPath.HasValue() == false)
```
Comment (ayende) [inline tools/Voron.Recovery/CommandLineApp.cs]: I don't want to create a server here. Instead, you should be creating an instance of `DocumentDatabase()` and use that API directly.

File: tools/Voron.Recovery/RecoveredDatabaseCreator.cs (hunk 5 above)
Comment (ayende) [reply tools/Voron.Recovery/RecoveredDatabaseCreator.cs]: But you aren't doing any disposal internally at all, no?

File: tools/Voron.Recovery/RecoveredDatabaseCreator.cs [hunk 14]
```diff
@@ -0,0 +1,433 @@
...
+        }
+
+        public void WriteRevision(Document document)
+        {
+            if (_database.DocumentsStorage.RevisionsStorage.Configuration == null)
+            {
+                var revisionsConfiguration = new RevisionsConfiguration {Default = new RevisionsCollectionConfiguration {Disabled = false}};
+                var configurationJson = EntityToBlittable.ConvertCommandToBlittable(revisionsConfiguration, _context);
+                (long index, _) = _database.ServerStore.ModifyDatabaseRevisions(_context, _database.Name, configurationJson, Guid.NewGuid().ToString()).Result;
+                AsyncHelpers.RunSync(() => _database.RachisLogIndexNotifications.WaitForIndexNotification(index, _database.ServerStore.Engine.OperationTimeout));
```
Comment (ayende) [inline tools/Voron.Recovery/RecoveredDatabaseCreator.cs]: You shouldn't be doing cluster operations here.

File: tools/Voron.Recovery/RecoveredDatabaseCreator.cs [hunk 15]
```diff
@@ -0,0 +1,433 @@
...
+
+        public void WriteDocument(Document document)
+        {
+            var actions = _databaseDestination.Documents();
+            WriteDocumentInternal(document, actions);
+        }
+
+        public void WriteRevision(Document document)
+        {
+            if (_database.DocumentsStorage.RevisionsStorage.Configuration == null)
```
Comment (ayende) [inline tools/Voron.Recovery/RecoveredDatabaseCreator.cs]: why are you dealing with the configuration in this manner?
If there are revisions in the DB, we want to preserve them.

It does bring a good point. Recovery won't help us with overall config (etl, revisions) or cluster wide items such as identity or compare exchange values.
Need to be sure that we make a note of this

File: tools/Voron.Recovery/RecoveredDatabaseCreator.cs [hunk 16]
```diff
@@ -0,0 +1,433 @@
...
+            // store this attachment either under relevant orphan doc, or under already seen doc
+            using (var tx = _context.OpenWriteTransaction())
+            {
+                var orphanAttachmentDocId = GetOrphanAttachmentDocId(hash);
+                var orphanAttachmentDoc = _database.DocumentsStorage.Get(_context, orphanAttachmentDocId);
+                if (orphanAttachmentDoc != null)
+                {
+                    // check which documents already seen for this attachment and attach it to them
+                    orphanAttachmentDoc.Data.Modifications = new DynamicJsonValue(orphanAttachmentDoc.Data);
+                    foreach (var docId in orphanAttachmentDoc.Data.GetPropertyNames())
```
Comment (ayende) [inline tools/Voron.Recovery/RecoveredDatabaseCreator.cs]: Let's use a real object here. That would make the code a lot easier and maintainable over time

File: tools/Voron.Recovery/RecoveredDatabaseCreator.cs [hunk 17]
```diff
@@ -0,0 +1,433 @@
...
+                            _database.DocumentsStorage.AttachmentsStorage.PutAttachment(_context, docId, originalName, originalContentType, hash, null,
+                                attachmentStream);
+                        }
+                    }
+
+                    using (var newDocument = _context.ReadObject(orphanAttachmentDoc.Data, orphanAttachmentDoc.Id,
+                        BlittableJsonDocumentBuilder.UsageMode.ToDisk))
+                    {
+
+                        if (newDocument.GetPropertyNames().Length == 1) // 1 for @metadata and @collection
```
Comment (ayende) [inline tools/Voron.Recovery/RecoveredDatabaseCreator.cs]:
```suggestion
                        if (newDocument.GetPropertyNames().Length == 1) // 1 for @metadata
```

File: tools/Voron.Recovery/RecoveredDatabaseCreator.cs [hunk 18]
```diff
@@ -0,0 +1,433 @@
...
+                    }
+                }
+                else
+                {
+                    VoronStream existingStream = null;
+                    using (var tree = _context.Transaction.InnerTransaction.CreateTree(_attachmentsSlice))
+                    {
+                        existingStream = tree.ReadStream(hash);
+                    }
+                    // although no previous doc asked for this attachment, it still might appear from previous recovery sessions. If so - ignore this WriteAttachment call
```
Comment (ayende) [inline tools/Voron.Recovery/RecoveredDatabaseCreator.cs]: What does previous recovery session mean?

File: tools/Voron.Recovery/RecoveredDatabaseCreator.cs [hunk 19]
```diff
@@ -0,0 +1,433 @@
...
+            }
+
+            var item = new DocumentItem {Document = document};
+            actions.WriteDocument(item, _results.Documents);
+            actions.Dispose();
+
+            if (hadCountersFlag)
+            {
+                List<CounterGroupDetail> counterGroupDetailCloneList = null;
+                using (var tx = _context.OpenWriteTransaction())
```
Comment (ayende) [inline tools/Voron.Recovery/RecoveredDatabaseCreator.cs]: why multiple separate transactions?

File: tools/Voron.Recovery/RecoveredDatabaseCreator.cs [hunk 20]
```diff
@@ -0,0 +1,433 @@
...
+        {
+            if (_logger.IsOperationsEnabled)
+                _logger.Operations(msg, ex);
+
+            DocumentsTransaction tx = null;
+            if (_context.HasTransaction == false)
+                tx = _context.OpenWriteTransaction();
+            try
+            {
+                var logDoc = _database.DocumentsStorage.Get(_context, _logDocId);
```
Comment (ayende) [inline tools/Voron.Recovery/RecoveredDatabaseCreator.cs]: This document is going to get HUGE, no?

File: tools/Voron.Recovery/Recovery.cs [hunk 21]
```diff
@@ -486,18 +556,23 @@ void PrintRecoveryProgress(long startOffset, byte* mem, byte* eof, DateTime now)
                             var message =
                                 $"Unexpected exception at position {GetFilePosition(startOffset, mem)}:{Environment.NewLine} {e}";
                             mem = PrintErrorAndAdvanceMem(message, mem);
+                            try
+                            {
+                                recoveredTool.Log(message, e);
+                            }
+                            catch (Exception)
+                            {
```
Comment (ayende) [inline tools/Voron.Recovery/Recovery.cs]: What scenario would cause it to throw here?

File: tools/Voron.Recovery/Recovery.cs [hunk 22]
```diff
@@ -685,33 +741,23 @@ private void ReportOrphanAttachmentsAndMissingAttachments(TextWriter writer, Bli
...
                 return;
             }
+
             if (_documentsAttachments.Count == 0)
             {
-                foreach (var (hash, tag, size) in _attachmentsHashs)
-                {
-                    ReportOrphanAttachmentDocumentId(hash, size, tag, documentsWriter);
-                }
+                foreach ((string hash, string tag, long size) in _attachmentsHashs) ReportOrphanAttachmentDocumentId(hash, size, tag, documentsWriter);
```
Comment (ayende) [inline tools/Voron.Recovery/Recovery.cs]: use {} here.

File: tools/Voron.Recovery/Recovery.cs [hunk 23]
```diff
@@ -1038,22 +1055,25 @@ private bool WriteCounter(byte* mem, int sizeInBytes, BlittableJsonTextWriter co
...
-                    [nameof(CounterItem.Batch.Values)] = counterGroup.Values
-                });
+                // context.Write(countersWriter,
+                //     new DynamicJsonValue
+                //     {
+                //         [nameof(CounterItem.DocId)] = counterGroup.DocumentId.ToString(),
+                //         [nameof(CounterItem.ChangeVector)] = counterGroup.ChangeVector.ToString(),
+                //         [nameof(CounterItem.Batch.Values)] = counterGroup.Values
+                //     });
+
```
Comment (ayende) [inline tools/Voron.Recovery/Recovery.cs]: why is this commented?

File: tools/Voron.Recovery/RecoveredDatabaseCreator.cs (hunk 5 above)
Comment (aviviadi) [reply tools/Voron.Recovery/RecoveredDatabaseCreator.cs]: I do: https://github.com/ravendb/ravendb/pull/10266/files#diff-54908524c9c84febb0b2eaa52e9d26c1R277

File: tools/Voron.Recovery/Recovery.cs (hunk 23 above)
Comment (aviviadi) [reply tools/Voron.Recovery/Recovery.cs]: this is still draft PR

File: tools/Voron.Recovery/RecoveredDatabaseCreator.cs (hunk 15 above)
Comment (aviviadi) [reply tools/Voron.Recovery/RecoveredDatabaseCreator.cs]: The recovered database is a new db. I am creating new config. Is there another way of doing that?

Comment (ayende) [reply tools/Voron.Recovery/RecoveredDatabaseCreator.cs]: This is again about not running a server. You can manually set this value to the right parameters, in memory only.

File: tools/Voron.Recovery/CommandLineApp.cs (hunk 13 above)
Comment (aviviadi) [reply tools/Voron.Recovery/CommandLineApp.cs]: Please elaborate, how can I create DocumentDatabase instance without ServerStore? (and therefor a RavenServer) without refactoring DocumentDatabase?

File: tools/Voron.Recovery/RecoveredDatabaseCreator.cs (hunk 20 above)
Comment (aviviadi) [reply tools/Voron.Recovery/RecoveredDatabaseCreator.cs]: No, logging just errors

File: tools/Voron.Recovery/Recovery.cs (hunk 21 above)
Comment (aviviadi) [reply tools/Voron.Recovery/Recovery.cs]: Some exception in the creation of the log document itself

File: tools/Voron.Recovery/RecoveredDatabaseCreator.cs (hunk 6 above)
Comment (aviviadi) [reply tools/Voron.Recovery/RecoveredDatabaseCreator.cs]: see https://github.com/ravendb/ravendb/pull/10266#discussion_r384347010

File: test/SlowTests/Issues/RavenDB-8451.cs (hunk 1 above)
Comment (aviviadi) [reply test/SlowTests/Issues/RavenDB-8451.cs:59]: EnsureDatabaseDeletion added (a bit later after opening new store for it)

File: test/SlowTests/Issues/RavenDB_11904.cs (hunk 2 above)
Comment (aviviadi) [reply test/SlowTests/Issues/RavenDB_11904.cs:31]: EnsureDatabaseDeletion added (a bit later after opening new store for it)

File: test/SlowTests/Issues/RavenDB_11905.cs (hunk 3 above)
Comment (aviviadi) [reply test/SlowTests/Issues/RavenDB_11905.cs:28]: EnsureDatabaseDeletion added (a bit later after opening new store for it)

File: test/SlowTests/Voron/RecoveryTest.cs (hunk 4 above)
Comment (aviviadi) [reply test/SlowTests/Voron/RecoveryTest.cs]: EnsureDatabaseDeletion added (a bit later after opening new store for it)

File: tools/Voron.Recovery/RecoveredDatabaseCreator.cs (hunk 9 above)
Comment (aviviadi) [reply tools/Voron.Recovery/RecoveredDatabaseCreator.cs:61]: I am not following? ..

File: test/SlowTests/Issues/RavenDB-8451.cs [hunk 24]
```diff
@@ -65,92 +67,38 @@ public async Task CanRecoverEncryptedDatabaseInternal(bool nullifyMasterKey = fa
...
-                DataFileDirectory = dbPath,
-                PathToDataFile = Path.Combine(dbPath, "Raven.voron"),
-                OutputFileName = Path.Combine(recoveryExportPath, "recovery.ravendump"),
-                MasterKey = masterKey,
-                DisableCopyOnWriteMode = nullifyMasterKey
-            }))
-            {
-                recovery.Execute(TextWriter.Null, CancellationToken.None);
-            }
+                var _ = store.Maintenance.Server.Send(new CreateDatabaseOperation(new DatabaseRecord {DatabaseName = recoverDbName}));
```
Comment (ppekrol) [inline test/SlowTests/Issues/RavenDB-8451.cs]: You do not need to declare variable

File: test/SlowTests/Issues/RavenDB-8451.cs [hunk 25]
```diff
@@ -65,92 +67,38 @@ public async Task CanRecoverEncryptedDatabaseInternal(bool nullifyMasterKey = fa
...
+                    recovery.Execute(TextWriter.Null, CancellationToken.None);
+                }
+            }
+
             using (var store = GetDocumentStore(new Options()
             {
                 AdminCertificate = certificates.ServerCertificate.Value,
                 ClientCertificate = certificates.ServerCertificate.Value,
             }))
+            using (var __ = EnsureDatabaseDeletion(recoverDbName, store))
```
Comment (ppekrol) [inline test/SlowTests/Issues/RavenDB-8451.cs]: You do not need to declare variable

File: test/SlowTests/Issues/RavenDB_11904.cs [hunk 26]
```diff
@@ -27,29 +29,34 @@ public async Task CanLoadDatabaseAfterUsingVoronRecoveryOnItWithCopyOnWriteMode(
...
-            using (var recovery = new Recovery(new VoronRecoveryConfiguration()
-            {
-                LoggingMode = Sparrow.Logging.LogMode.None,
-                DataFileDirectory = dbPath,
-                PathToDataFile = Path.Combine(dbPath, "Raven.voron"),
-                OutputFileName = Path.Combine(recoveryExportPath, "recovery.ravendump"),
-            }))
-            {
-                recovery.Execute(TextWriter.Null, CancellationToken.None);
+                var _ = store.Maintenance.Server.Send(new CreateDatabaseOperation(new DatabaseRecord {DatabaseName = recoverDbName}));
```
Comment (ppekrol) [inline test/SlowTests/Issues/RavenDB_11904.cs]: You do not need to declare variable

File: test/SlowTests/Issues/RavenDB_11904.cs [hunk 27]
```diff
@@ -73,19 +80,12 @@ public async Task CanLoadDatabaseAfterUsingVoronRecoveryOnItWithCopyOnWriteMode(
             // let's import the recovery files
 
             using (var store = GetDocumentStore())
+            using (var __ = EnsureDatabaseDeletion(recoverDbName, store))
```
Comment (ppekrol) [inline test/SlowTests/Issues/RavenDB_11904.cs]: You do not need to declare variable

File: test/SlowTests/Issues/RavenDB_11905.cs [hunk 28]
```diff
@@ -23,35 +25,32 @@ public async Task CanUseVoronRecoveryOnEmptyDatabase()
...
+
+            using (var store = GetDocumentStore(new Options()
             {
                 Path = dbPath
             }))
             {
-                
+                var _ = store.Maintenance.Server.Send(new CreateDatabaseOperation(new DatabaseRecord {DatabaseName = recoverDbName}));
```
Comment (ppekrol) [inline test/SlowTests/Issues/RavenDB_11905.cs]: You do not need to declare variable

File: test/SlowTests/Issues/RavenDB_11905.cs [hunk 29]
```diff
@@ -23,35 +25,32 @@ public async Task CanUseVoronRecoveryOnEmptyDatabase()
...
-                OutputFileName = Path.Combine(recoveryExportPath, "recovery.ravendump"),
+                LoggingOutputPath = Path.Combine(recoveryExportPath),
+                RecoveredDatabase = recoveredDatabase
             }))
             {
                 recovery.Execute(TextWriter.Null, CancellationToken.None);
             }
             
             using (var store = GetDocumentStore())
+            using (var __ = EnsureDatabaseDeletion(recoverDbName, store))
```
Comment (ppekrol) [inline test/SlowTests/Issues/RavenDB_11905.cs]: You do not need to declare variable

File: test/SlowTests/Voron/RecoveryTest.cs (hunk 4 above)
Comment (ppekrol) [reply test/SlowTests/Voron/RecoveryTest.cs]: You do not need to declare variable

File: test/SlowTests/Voron/RecoveryTest.cs [hunk 30]
```diff
@@ -0,0 +1,277 @@
...
+            using var x = EnsureDatabaseDeletion(recoverDbName, store);
```
Comment (ppekrol) [inline test/SlowTests/Voron/RecoveryTest.cs]: You do not need to declare variable

File: tools/Voron.Recovery/RecoveredDatabaseCreator.cs (hunk 9 above)
Comment (ppekrol) [reply tools/Voron.Recovery/RecoveredDatabaseCreator.cs:61]: Even in local function. So someone will know what you are doing here.

File: tools/Voron.Recovery/RecoveredDatabaseCreator.cs (hunk 15 above)
Comment (ppekrol) [reply tools/Voron.Recovery/RecoveredDatabaseCreator.cs]: Different idea. Perform a recovery on a live server and take database down? Copy values from Server Storage (Cluster), etc?

File: tools/Voron.Recovery/RecoveredDatabaseCreator.cs (hunk 5 above)
Comment (ppekrol) [reply tools/Voron.Recovery/RecoveredDatabaseCreator.cs]: Why are you enforcing commit? What is the idea behind it?

File: tools/Voron.Recovery/RecoveredDatabaseCreator.cs [hunk 31]
```diff
@@ -0,0 +1,427 @@
...
+                var configurationJson = EntityToBlittable.ConvertCommandToBlittable(revisionsConfiguration, _context);
+                (long index, _) = _database.ServerStore.ModifyDatabaseRevisions(_context, _database.Name, configurationJson, Guid.NewGuid().ToString()).Result;
+                AsyncHelpers.RunSync(() => _database.RachisLogIndexNotifications.WaitForIndexNotification(index, _database.ServerStore.Engine.OperationTimeout));
+            }
+            var actions = _databaseDestination.RevisionDocuments();
+            WriteDocumentInternal(document, actions);
+        }
+
+        public void WriteCounterItem(CounterGroupDetail counterGroupDetail, bool skipDocExistence = false)
+        {
```
Comment (ppekrol) [inline tools/Voron.Recovery/RecoveredDatabaseCreator.cs]: Let's create a separate class to handle each one of those. At this moment it is hard to follow and maintain.

File: tools/Voron.Recovery/RecoveredDatabaseCreator.cs (hunk 5 above)
Comment (aviviadi) [reply tools/Voron.Recovery/RecoveredDatabaseCreator.cs]: To use different transactions.

I am using DatabaseDestination (and not direct PUT command) in order to be compatible with what import does for instance. And then I want to commit (`actions.Dispose()`) and add another transaction _only if needed_. 

I cannot do everything in a single transaction AND use DatabaseDestination/TxMerger.

File: tools/Voron.Recovery/RecoveredDatabaseCreator.cs (hunk 31 above)
Comment (aviviadi) [reply tools/Voron.Recovery/RecoveredDatabaseCreator.cs]: I am not following... WriteCounterItem, WriteDocument, WriteRevisions, WriteConflicts and WriteAttachments exists.
They are 99% of this class any how? (or do you mean to split something else to a different class?)

Comment (ppekrol) [reply tools/Voron.Recovery/RecoveredDatabaseCreator.cs]: Counter actions wrap in CountersRecover, Document actions wrap in DocumentRecovery, etc. A bit of OOP haven't hurt anyone.

Comment (aviviadi) [reply tools/Voron.Recovery/RecoveredDatabaseCreator.cs]: Done here : https://github.com/ravendb/ravendb/pull/10266/commits/15db3036b2bbecff6f08aa46fdc062efb8ae55f6

I am not so sure it is needed though.. In my eyes it brings more reading complexity.. 
And there is no real _pattern_ here.  (unless we want to revise Recovery.cs which I guess we do not)

File: test/SlowTests/Issues/RavenDB-8451.cs [hunk 32]
```diff
@@ -65,92 +67,37 @@ public async Task CanRecoverEncryptedDatabaseInternal(bool nullifyMasterKey = fa
...
+                    recovery.Execute(TextWriter.Null, CancellationToken.None);
+                }
+            }
 
             using (var store = GetDocumentStore(new Options()
             {
                 AdminCertificate = certificates.ServerCertificate.Value,
                 ClientCertificate = certificates.ServerCertificate.Value,
             }))
+            using (EnsureDatabaseDeletion(recoverDbName, store))
```
Comment (ppekrol) [inline test/SlowTests/Issues/RavenDB-8451.cs:95]: After recovery let's deploy all indexes from sample data (you can export them in the test) and check if the counts on valid data and after recovery are the same + also if there are no indexing errors

File: test/SlowTests/Issues/RavenDB_11904.cs [hunk 33]
```diff
@@ -73,19 +79,12 @@ public async Task CanLoadDatabaseAfterUsingVoronRecoveryOnItWithCopyOnWriteMode(
             // let's import the recovery files
 
             using (var store = GetDocumentStore())
+            using (EnsureDatabaseDeletion(recoverDbName, store))
```
Comment (ppekrol) [inline test/SlowTests/Issues/RavenDB_11904.cs:82]: After recovery let's deploy all indexes from sample data (you can export them in the test) and check if the counts on valid data and after recovery are the same + also if there are no indexing errors

File: tools/Voron.Recovery/CommandLineApp.cs [hunk 34]
```diff
@@ -67,16 +80,42 @@ private static void ConfigureRecoveryCommand()
...
                 var ignoreDataIntegrityErrorsOfAlreadySyncedTransactionsArg = cmd.Option("--IgnoreInvalidDataErrorsOfAlreadySyncedTransactions", "Default is false.", CommandOptionType.SingleValue);
                 var ignoreInvalidPagesInARowArg = cmd.Option("--IgnoreInvalidPagesInARow", "Default is false.", CommandOptionType.SingleValue);
-
                 var loggingModeArg = cmd.Option("--LoggingMode", "Logging mode: Operations or Information.", CommandOptionType.SingleValue);
-
                 var masterKey = cmd.Option("--MasterKey", "Encryption key: base64 string of the encryption master key", CommandOptionType.SingleValue);
 
                 cmd.OnExecute(() =>
                 {
+                    var recoveredDatabaseName = "RecoveredDB-" + Guid.NewGuid();
```
Comment (ppekrol) [inline tools/Voron.Recovery/CommandLineApp.cs:88]: let's put date in the name

File: tools/Voron.Recovery/CommandLineApp.cs [hunk 35]
```diff
@@ -67,16 +80,42 @@ private static void ConfigureRecoveryCommand()
...
+                    var serverConfig = RavenConfiguration.CreateForServer("RecoveryConfiguration");
+                    serverConfig.Core.FeaturesAvailability = FeaturesAvailability.Experimental;
+                    serverConfig.Licensing.EulaAccepted = true;
+                    serverConfig.Core.DataDirectory = new PathSetting(recoverDirectoryArg.Value);
+                    serverConfig.Initialize();
+
+                    LicenseManager.IgnoreProcessorAffinityChanges = true;
```
Comment (ppekrol) [inline tools/Voron.Recovery/CommandLineApp.cs:95]: this is for testing purposes only, you should remove this

File: tools/Voron.Recovery/RecoveredDatabaseCreator.cs (hunk 15 above)
Comment (ppekrol) [reply tools/Voron.Recovery/RecoveredDatabaseCreator.cs]: Ping
